### DISCUSSION TITLE
Error trying to compute altaz from earth position to jupiter

### DISCUSSION THREAD

**Discussion by sizzzzlerz** [skyfielders/python-skyfield#1008] [General]
My ultimate goal is to compute the angles between the largest moons of Jupiter and Jupiter. To facilitate that, I loaded the  jup365.bsp  Spice kernel which includes Jupiter, the sun and Earth and  a number of moons. To familiarize myself with this file, I first tried computing the alt and az  from my location (home) to jupiter using home.at(t).observe(jupiter).apparent().altaz().  What I got was a KeyError exception that basically stated that Saturn was not found in the kernel.  Why does Saturn need to be in the kernel? Is this a possible bug?

**Comment by Bernmeister** [comment]
Although I cannot offer specific help, I'm confident the first thing those that _can_ help you will ask for a small complete test script which demonstrates the error.

**Reply by sizzzzlerz** [reply]
Here is a script that demonstrates what I'm seeing. Note that the jup365.bsp is over 1GB so if you don't already have it, it is going to take some time to download.

```
from skyfield.api import load, wgs84

ts = load.timescale()
t = ts.now()

# change the path here as appropriate
eph = load("/Users/lskea/Downloads/jup365.bsp")
print(eph)

earth = eph['earth']
jupiter = eph['jupiter']

astrometric = earth.at(t).observe(jupiter)

ra,dec,dist = astrometric.radec()
print(ra, dec, dist)

altaz = astrometric.apparent().altaz()

```

**Comment by labonocl** [comment]
Hei,

when calling `apparent()` on a position, according to the docs the following happens:

> This applies two effects to the position that arise from
> relativity and shift slightly where the other body will appear
> in the sky: the deflection that the image will experience if its
> light passes close to large masses in the Solar System, and the
> aberration of light caused by the observer's own velocity.

And indeed, the code always applies the deflection of light due to the Sun, Jupiter, and Saturn. In relativity.py it defines deflectors

`deflectors = ['sun', 'jupiter', 'saturn', 'moon', 'venus', 'uranus', 'neptune']`

And the `add_deflection()` function actually has a keyword called `count`, that is set to 3 as default, i.e., the first three elements (Sun, Jupiter, Saturn) are included. The function `apparent()` currently has no way of controlling the `count` keyword.

So I see a few solutions:

- Load Saturn's ephemeris also
- change your local skyfield code to either set the `count` default value to 2 (in `add_deflection()` in relativity.py), or add a `count` keyword to `apparent()` (in positionlib.py) which passes it on to `add_deflection()`

From a skyfield development point-of-view, issuing a warning when a deflection cannot be applied due to lacking ephemeris is probably the cleanest way.

BTW: as it is now, your code will not work, as altaz() needs to be called from a position on Earth (i.e. a latitude/longitude). But that is easily fixed. Also, it is the `apparent()` call that causes the error, not the `altaz()` call.

**Reply by labonocl** [reply]
Regarding my first point, the following works:

```
from skyfield.api import load, wgs84, N, W

ts = load.timescale()
t = ts.now()

# change the path here as appropriate
eph = load("jup365.bsp")
bodies = load('de421.bsp')
# push Saturn's ephemeris in there - not sure how safe this is
eph.spk.segments.append(bodies.spk.segments[5])
eph.segments.append(bodies.segments[5])
eph.codes.add(6)

print(eph)

earth = eph['earth']
jupiter = eph['jupiter']

boston = earth + wgs84.latlon(42.3583 * N, 71.0603 * W, elevation_m=43)
astrometric = boston.at(t).observe(jupiter)

ra,dec,dist = astrometric.radec()
print(ra, dec, dist)

altaz = astrometric.apparent().altaz()
print(altaz)
```

BTW, I also calculated the position without Saturn's contribution to the deflection, and (as expected?) in this case the difference in angle is of the order of 1e-8 degrees.

**Reply by sizzzzlerz** [reply]
I caught the mistake about calling altaz  on apparent after I'd hurriedly typed in and submitted here what I'd been running in iPython which did correctly make the call to apparent. I just missed that line when searching for the lines to enter into the script. Thanks for catching here however.

Maybe you or someone else can address why perturbations from Saturn would affect light traveling to Jupiter. Won't Jupiter always be in front of Saturn or am I missing some scenario where it would have an affect?

**Reply by labonocl** [reply]
![Screenshot 2024-10-07 114819](https://github.com/user-attachments/assets/cd4909d5-d994-4ab8-adb1-4d9266655ef4)
The image above sketches a kind of worst case scenario. The orbits of the three planets Earth (E), Jupiter (J), Saturn (S) are to scale, the planet sizes obvisouly not. Observing Jupiter from Earth (without Saturn), the light would travel the dotted line. However, because of Saturn's mass, the light travels along a curved path (highly exaggerated in this schematic), such that you would assume Jupiter is where the dashed line points.
However, two comments:

1. According to your original post, you are interested in computing "the angles between the largest moons of Jupiter and Jupiter". Since the moons are very close to Jupiter (in comparisson to the distance to Saturn), the light paths from the moons and Jupiter itself will be affected by Saturn in essentially the same way, such that the angles between moons and Jupiter will likely not be affected (much).
2. Obviously it all depends on your requirements, but putting in some numbers for the expected deflection in the case sketched above, I get the deflection angle to be about 1 micro arcsecond, which is (well) below the angular resolution of current instruments. But again, it all depends on the accuracy that you require for your application.

**Reply by sizzzzlerz** [reply]
Thanks for your explanation. My requirements are I  just want to draw pretty figures of the movement of these moons around Jupiter like you find in various astronomy magazines. I really don't think a 1 microsecond discrepancy is going to significantly alter them 😁
